### PR TITLE
OneDNN md-in-tensor refactoring part 5: Memory descriptor enabled for elementwises, reductions and expand_v2 ops

### PR DESCRIPTION
**Pull request by jakpiase**
### PR types
Others

### PR changes
OPs

### Describe
Memory descriptor enabled for elementwises, reductions and expand_v2 ops

### CODE REVIEW

**Comment by jakpiase** [comment]
Reopened due to license/cla being stuck

**Review by jczaja — DISMISSED** [review]
LGTM

**Comment by jakpiase** [comment]
@tsocha please review this PR

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 1]
```diff
@@ -718,34 +706,28 @@ class BroadcastDataMKLDNNHandler
  public:
   BroadcastDataMKLDNNHandler(const dnnl::algorithm algo,
                              const dnnl::engine engine,
-                             platform::Place cpu_place, const Tensor* out,
-                             const Tensor* x, float scale_x, float scale_y,
-                             const dnnl::memory::desc& x_mem_desc)
+                             platform::Place cpu_place, const Tensor* x,
+                             Tensor* out, float scale_x, float scale_y,
+                             const std::vector<int64_t>& extended_x_dims)
```
Comment (tsocha) [inline paddle/fluid/platform/mkldnn_reuse.h:711]: [CONCERN]
Type is almost the same but order has changed.
I hope that our tests will detect potential problems with it.

Was this change of order necessary?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 2]
```diff
@@ -718,34 +706,28 @@ class BroadcastDataMKLDNNHandler
...
+    const auto src1_md = x->mem_desc().reshape(extended_x_dims);
 
-    const auto src1_md = x_mem_desc;
-
-    dnnl::primitive_attr attributes;
-    attributes.set_scales(DNNL_ARG_SRC_0, 0, {scale_x});
-    attributes.set_scales(DNNL_ARG_SRC_1, 0, {scale_y});
+    dnnl::primitive_attr attrs;
+    attrs.set_scales(DNNL_ARG_SRC_0, 0, {scale_x});
+    attrs.set_scales(DNNL_ARG_SRC_1, 0, {scale_y});
```
Comment (tsocha) [inline paddle/fluid/platform/mkldnn_reuse.h]: [OPINION]
The old name was good.

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 3]
```diff
@@ -758,22 +740,18 @@ class ReductionMKLDNNHandler
   ReductionMKLDNNHandler(const dnnl::algorithm algo, const float p,
                          const float eps, const dnnl::engine engine,
                          platform::Place cpu_place, const Tensor* x,
-                         const Tensor* y, std::vector<int64_t> y_tz,
-                         const dnnl::primitive_attr& attr = NULL)
+                         const Tensor* out, std::vector<int64_t> out_tz,
+                         const dnnl::primitive_attr& attrs = NULL)
```
Comment (tsocha) [inline paddle/fluid/platform/mkldnn_reuse.h:744]: [COMMENT]
If `dnnl::primitive_attr` can hold multiple attributes its name should be changed in onednn, but if it can't then our parameter should have the old name.

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h [hunk 4]
```diff
@@ -29,11 +29,11 @@ inline std::vector<int64_t> CalculateReducedDims(
     bool reduce_all, bool keep_dim) {
   if (keep_dim) return phi::vectorize(output->dims());
 
-  if (reduce_all)
-    return std::vector<int64_t>(phi::vectorize(input->dims()).size(), 1);
+  if (reduce_all) return std::vector<int64_t>(input->dims().size(), 1);
 
-  std::vector<int64_t> output_dims(phi::vectorize(input->dims()));
+  std::vector<int64_t> output_dims = phi::vectorize(input->dims());
```
Comment (tsocha) [inline paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]: [QUESTION]
Was this change necessary?

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h [hunk 5]
```diff
@@ -29,11 +29,11 @@ inline std::vector<int64_t> CalculateReducedDims(
...
   for (size_t i = 0; i < reduce_dims.size(); ++i) {
+    // handle negative dims, f.e. -1 means last dimension
```
Comment (tsocha) [inline paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]: [CONCERN]
I think `last dimension` is not the best description.
`(1, 3, -1, 512, 512)` << the last dimension is `-1` or `512`?

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h [hunk 6]
```diff
@@ -52,51 +52,52 @@ class ReduceMKLDNNKernel : public framework::OpKernel<T> {
...
     auto reduce_dims = ctx.Attr<std::vector<int>>("dim");
     bool reduce_all = ctx.Attr<bool>("reduce_all");
     bool keep_dim = ctx.Attr<bool>("keep_dim");
 
-    auto output_dims =
-        CalculateReducedDims(input, output, reduce_dims, reduce_all, keep_dim);
-    auto input_dims = phi::vectorize(input->dims());
+    auto x_tz = phi::vectorize(x->dims());
+    auto out_tz =
+        CalculateReducedDims(x, out, reduce_dims, reduce_all, keep_dim);
```
Comment (tsocha) [inline paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h:64]: [OPINION]
I see why you changed names here but `input`>`x` IMO

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h [hunk 7]
```diff
@@ -52,51 +52,52 @@ class ReduceMKLDNNKernel : public framework::OpKernel<T> {
...
+          x_tz, framework::TransToProtoVarType(x->dtype()), x_type,
+          onednn_engine);
 
       auto reorder_src_memory_p = reorder_handler.AcquireSrcMemory(
-          input->mem_desc(), platform::to_void_cast(input->data<T>()));
+          x->mem_desc(), platform::to_void_cast(x->data<T>()));
 
-      auto reorder_dst_memory_p = reorder_handler.AcquireDstMemory(
-          output, input->mem_desc(), ctx.GetPlace());
+      // reuse same mem desc since it is a simple copy
```
Comment (tsocha) [inline paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]:
```suggestion
      // reuse mem desc since it is a simple copy
```

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h [hunk 8]
```diff
@@ -124,25 +126,18 @@ class ReduceGradMKLDNNKernel : public framework::OpKernel<T> {
     bool keep_dim = ctx.Attr<bool>("keep_dim");
     bool reduce_all = ctx.Attr<bool>("reduce_all");
     auto dims = ctx.Attr<std::vector<int>>("dim");
-    const auto* dout = ctx.Input<Tensor>(framework::GradVarName("Out"));
+    auto* dout = ctx.Input<Tensor>(framework::GradVarName("Out"));
```
Comment (tsocha) [inline paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]: [CONCERN]
Do you plan to change this pointer?

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h [hunk 9]
```diff
@@ -124,25 +126,18 @@ class ReduceGradMKLDNNKernel : public framework::OpKernel<T> {
...
-        CalculateReducedDims(dx, dout, dims, reduce_all, keep_dim);
-    const auto output_dims = phi::vectorize(dx->dims());
-
-    auto dout_mem_desc = dout->mem_desc();
-
-    if (input_dims != output_dims) {
-      dout_mem_desc = dout_mem_desc.reshape(input_dims);
-    }
+    auto dout_tz = CalculateReducedDims(dx, dout, dims, reduce_all, keep_dim);
+    auto dx_tz = phi::vectorize(dx->dims());
```
Comment (tsocha) [inline paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h:133]: [CONCERN]
The logic changed here.
Input has been swaped with output, was it intentional?

What means `tz` and `d` in this context?

**Review by tsocha — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h (hunk 4 above)
Comment (jakpiase) [reply paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]: No, it will produce the exact same assembly, I will revert it then

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h (hunk 5 above)
Comment (jakpiase) [reply paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]: You're right, I'll change it to "rightmost"

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h (hunk 6 above)
Comment (jakpiase) [reply paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h:64]: I have changed them to unify all of our ops, since all new ops are following this pattern, i.e. inputs are described as 'X' and 'Y' and output is described as 'Out'. Some legacy operators are having different names because of backwards compatibility, but I was thinking that unifying them to get rid of the additional noise would be nice

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h (hunk 8 above)
Comment (jakpiase) [reply paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h]: Nope, nice catch, that was a mistake, thanks

File: paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h (hunk 9 above)
Comment (jakpiase) [reply paddle/fluid/operators/reduce_ops/mkldnn/reduce_mkldnn_op.h:133]: `Input` and `Output` were incorrectly swapped here earlier. Reduce kernel was one of the first ops that I have made and there were some inconsistencies and bugs inside of it, so that change is intentional

`d` means that it is a grad tensor, so `dout` is the input and `dx` is the output
I have no idea what `tz` means, but it is commonly used as a synonym for `dims` inside PaddlePaddle, oneDNN and other frameworks

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 1 above)
Comment (jakpiase) [reply paddle/fluid/platform/mkldnn_reuse.h:711]: It was not necessary, but for me order: input, output is more intuitive than output, input. I have checked all the tests for reduce_grad and expand_v2 and they all work correctly after my changes

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (jakpiase) [reply paddle/fluid/platform/mkldnn_reuse.h:744]: It can hold scales, post-ops, zero-points, so it definitely supports multiple attributes, but probably because of backward compatibility it won't be changed inside oneDNN

**Review by tsocha — APPROVED** [review]

**Review by jczaja — APPROVED** [review]
LGTM
